### ISSUE TITLE
Suggestion: Set absolute size for TextVisual instead of font_size

### ISSUE THREAD

**Issue by jdreaver** [vispy/vispy#1072]
I am trying to make a football field in vispy for a project I am working on. I tried to use a TextVisual to make the numbers, but I have one problem: the text resizes as the viewbox resizes. Is it possible to set an absolute size for TextVisuals instead of `font_size`? For example, if I set `size=(4, 6)`, the text would stay in a 4 by 6 window in the absolute coordinates of the scene. Currently, zooming out keeps the *screen size* of the text constant, but in my example that would make the yard line numbers shrink and grow :)

Sorry if this is a duplicate of another issue, I tried searching the repo for it.

**Comment by campagnola** [comment]
The way this is supposed to work is that the text visual has a certain font size, and it consults its "document" coordinate system to determine how to interpret that font size. By simply assigning a different  document coordinate system you should be able to control the font size however you like. We should also expose a simple API switch that would let the visual act as its own document, which will cover most use cases for this feature.

That said, I am not sure anyone has tried to do this yet.

**Comment by Eric89GXL** [comment]
Not a duplicate to my knowledge, but I think there are related issues.

I think we need a general way of specifying what coordinate frame sizes and
positions are in. Right now text size is in points, which is related to
logical pixels by DPI, while text position is in visual / document (I think
that's what it's called) coordinates. Then some other sizes like line
widths are in logical pixels.

I wonder if we can take care of this in some general way. It would be very
useful. @campagnola have you thought of a system for this problem? I guess
we could add an argument for each variable for the units/coordinate system
for the relevant variables, and then do something general at the Visual
level to deal with it.

**Comment by Eric89GXL** [comment]
@campagnola I think that solution would require e.g. updating the document
system with zoom, and doesn't seem as clean as specifying the desired units
of variables.

**Comment by campagnola** [comment]
Not sure what you mean. By default, all visuals use the scene's document coordinate system as their own. That means if you specify a font size of 12 pixels, then it will render as 12 pixels relative to the scene's document CS. If you make the text visual its own document, then the text would render as 12-units relative to the visual's coordinate system, and it would automatically scale as the visual's view was zoomed. Or if you create a "page" visual and make that the document cs of a text visual, then the text should always appear fixed to an exact area of the page, regardless of how it is zoomed.

**Comment by Eric89GXL** [comment]
For example being able to set marker sizes in visual coordinate units
instead of pixels would solve another user's issue, too. I can take a crack
at the API and implementation this week possibly if it sounds reasonable.

**Comment by Eric89GXL** [comment]
In my mind the units change is simpler for users, too.

**Comment by campagnola** [comment]
This need not be a part of the user-facing API; it's just the way I think it should be implemented. This is the entire reason for the concept of the document coordinate system--by making this something that we explicitly map to/from, it can be swapped out with any other coordinate system, which achieves the desired effect across all visuals (as long as they correctly reference that cs). It also takes advantage of the fact that scenegraph nodes inherit their document cs from their parent, which makes it possible to lay out entire figures in publication-ready coordinates, rather than in screen coordinates.

**Comment by Eric89GXL** [comment]
I'm not sure these "unit" types of changes I'm thinking of can generally be done by changing the coordinate system definitions, though. For example, if we have two different parameters that both use the `document` coordinate system at the GLSL level, and a user wants to specify only one of them in visual coordinates, then changing what the `document` coordinate system means will not work. At the GLSL level we need some general mechanism for getting data from one coordinate frame -- the one the user requests -- into another, which is the level that the GLSL code needs. So somewhere we need something like `$units_transform(size)` on each variable that takes it from the coordinate frame specified by the user (which they should be able to change with a property), into the proper coordinate system for GLSL. Does that makes sense?

If you agree we need these in GLSL to take care of the problem generally, then I don't see the point in changing what the `document` coordinate system is in this example -- I see it as a different `units_transform` being applied at the GLSL level. Instead of using the document coordinate system, it should use the visual one, in this case.

**Comment by Eric89GXL** [comment]
And I guess that should be `size_units_transform`, above, since you'd need one for each variable that could potentially have different units.

**Comment by Eric89GXL** [comment]
Even for text, I think I'd want to be able to say I want the position in document coords, and the size in visual coords. I don't think your coordinate modification idea would work for that case. If it does, it seems more opaque than a solution that had one transform for the position, and another for the size, each of which was set by a user-facing property for what units they are specifying the position and size in.

**Comment by campagnola** [comment]
> I don't see the point in changing what the document coordinate system is in this example -- I see it as a different units_transform being applied at the GLSL level. Instead of using the document coordinate system, it should use the visual one, in this case.

You've got me confused. What you are suggesting we do with `units_transform` is the entire _purpose_ of the document coordinate system. If you want a visual to have multiple document coordinate systems, then we could do that, and it will be just as complicated as having multiple `units_transform` per visual. 

However, I can't think of a real use case for this--perhaps you want text whose size is defined relative to its own CS, so it changes size as the view zooms, but you also want to draw an outline around the text that is always 1px on screen. Why would you want this? If you zoom out, then eventually all of the text area is covered by border (which is exactly the problem we had with the scaling markers implementation using this approach). 

> I think I'd want to be able to say I want the position in document coords, and the size in visual coords.

We already support this behavior. The reverse is exactly the way text operates now--its position is specified in scene coordinates and the size in document coordinates.

**Comment by campagnola** [comment]
(correction: position is specified in visual coordinates)

**Comment by Eric89GXL** [comment]
I don't think what I'm proposing is redundant with what already exists, but rather extends it. Right now, the relationship between a given input parameter (`size`, `position`, etc.) and the coordinate system that parameter *refers to* / maps into is fixed. Your proposal, if I understand it correctly, is that modifying the *coordinate system* of the visual should be used to get the desired result. It seems more general and more intuitive to me (both at the user and developer/maintenance level) to be able to modify which coordinate frame your input data *refers to* instead. These in my mind are complementary, not overlapping ideas.

What I'm proposing is functionality along the lines of `transAxes` and so forth in matplotlib. MPL has different axes and figure coordinate frames, and input parameters to functions live in those spaces. But sometimes it's nice to be able to specify input parameters in a different coordinate frame, so `transAxes` gives you a way to get your input parameters from one to the other. I don't think that having different ways of specifying data in figure or axes coordinates is redundant in their case with having those two coordinate systems, so I'm not sure why it would be in our case.

Some potential difficulties that I see with the change-coordinate-definition instead of change-coordinate-referral/"units" are:

1. If we have two different input parameters that use the same coordinate system (e.g., the document coordinate system) by default, the input data units cannot be changed independently.
2. The `scene` system takes care of setting up the different coordinate systems with nice reasonable relationships / hierarchy so that changes propagate nicely. It's hard for me to see how the change-CF method would work here.
3. I think it's much easier for users to think about which CF they want their data to live in, than think about changing the CFs themselves to get their data to look right. If we agree users having a `size_units='document'` or something (for the Text example) is the right approach, then this would map directly onto the implementation as well, which lowers overhead for developers.

On the other hand, I don't really see drawbacks to this approach. It seems more general, and more straightforward to understand from a developer perspective.

> However, I can't think of a real use case for this

I know I have been working with `vispy` before and wanted the ability to change the units of a given input parameter, and not known how to. It's not hard for me to imagine that, in some of those cases or in future cases, other inputs will be defined (by default) to be in the same coordinate frame, and therefore can't be changed independently only by changing the CF definitions. If we ever add another parameter to `Text` that also by default is in Document units, will even become a case where independently specifying which CF input parameters refer to is useful.

> We already support this behavior. The reverse is exactly the way text operates now--its position is specified in scene coordinates and the size in document coordinates.

It might be useful if we both describe how the user API and underlying code would accomplish this. In my plan, the user API would be adding `size_units` and `position_units` parameters. In our GLSL, we'd add two extra variables that map the user-inputted `units` variables to the proper CFs. In our Python code, we'd add two extra `setters` that choose those variables via `tr_sys.get(...)` to be the proper transforms during the view transform update.

I'm not sure if you're saying if the user-facing API should be the same, but I think you're saying that somewhere, these two coordinate frame relationships would actually be swapped, so that the GLSL wouldn't need to be modified. I'm not really sure how this would work e.g. with a Visual in a viewbox in a scene, where all these transforms are automatically supplied, for instance.

**Comment by sofroniewn** [comment]
@djhoese do you know if there's been any further developments on this feature request? I couldn't find any other discussion on the vispy GitHub.

I just made https://github.com/napari/napari/issues/3185 over on the napari GitHub for us to track this, but it could be something we're interested in contributing upstream if that would be desirable for you too. Any thoughts/ perspective would appreciated before we start. Thanks!!

**Comment by djhoese** [comment]
Not that I know of. Looks like @alisterburt is interested in helping which would be good because this may take a lot of tinkering to figure out the best way. I'm not sure I followed (I didn't read it all) this conversation but as with most software decisions there is probably a quick way to get this to work and the "right" way that will involve larger scale changes to VisPy. That said, a lot has changed in VisPy since this issue was first created so maybe it is easier than it used to be.
